peps/pep-0667.rst

```diff
@@ -371,6 +371,66 @@ C API
 As with all functions that return a borrowed reference, care must be taken to
 ensure that the reference is not used beyond the lifetime of the object.
 
+Impact on ``exec()``
+====================
+
+Even though this PEP does not modify ``exec()`` directly, the semantic change
```

Comment:
Thanks for the first pass at trying to document this behavior. I would suggest reframing this a bit to:
- Call out the specific semantic change to `locals()` to give the reader context
- Explain concisely how this PEP impacts or explicitly documents why a behavior change may be seen when using `exec()` with performance optimized `locals().`
- Make clear the expected behavior before explaining the examples.

Reply:
For the semantic changes to `locals()`, it's mentioned in this PEP a bit earlier. We should assume the reader already reads that part right? How should I either refer back or summarize that?

The reason of the behavior change is basically

> The core difference is that the current implementation returns the same dict
when ``locals()`` is called multiple times in function scope, so the following
code magically works

About the expected behavior, do you mean something like "the local variable update will not be propagated between `exec()`"? I feel difficult to describe the exact behavior otherwise I would probably just summarize it in words instead of examples. This is a rather subtle change and the original behavior is kind of odd. I would love any help from people who are better at documentation.

Reply:
@willingc I feel that these concepts are pretty well communicated by the current draft, with my suggested changes incorporated. First the current behavior of `exec()` is explained (with an example), then the reason for that behavior (and the connection to `locals()` is explained, then some additional confusing implications of the current behavior are shown (with an example), and then the new behavior is explained (which doesn't take very long, since it's much simpler), and then examples are given to demonstrate how to accomplish some things that are currently possible, under the new behavior. Are there particular aspects of this presentation that you don't find sufficiently clear?

Reply:
Thanks @carljm. Let's accept your suggestions, and I will give it another read. I'm primarily concerned that the expected behavior for 3.13+ is clear to readers. That may be so after your changes are included. I have a hard time following the wording flow until I see it as one prose document.

Reply:
After my Discord comments about even PEP 558 not explaining the locals() changes particularly well, I went back to PEP 558, the related python-dev thread, and even PEP 3100 to find the material that actually went into the decision to make locals() and exec() work as PEP 558 and PEP 667 both propose.

The early iterations of PEP 558 actually tried to standardise ``locals()`` without changing it, so hopefully the new text makes it clear why that approach didn't last long :)

Reply:
I believe the current PR text covers @willingc's concerns (please review again if I'm mistaken about that)